compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala

```diff
@@ -0,0 +1,238 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import Constants.Constant
+import Contexts.Context
+import Decorators._
+import Flags._
+import ast.Trees._
+import Definitions._
+import DenotTransformers._
+import StdNames._
+import Symbols._
+import MegaPhase._
+import Types._
+import dotty.tools.dotc.ast.tpd
+
+import scala.annotation.tailrec
+
+/** Optimize generic operations on tuples */
+class TupleOptimizations extends MiniPhase with IdentityDenotTransformer {
+  import tpd._
+
+  def phaseName: String = "genericTuples"
+
+  override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree = {
+    if (tree.symbol == defn.DynamicTuple_dynamicCons) transformTupleCons(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicTail) transformTupleTail(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicSize) transformTupleSize(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicConcat) transformTupleConcat(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicApply) transformTupleApply(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicToArray) transformTupleToArray(tree)
+    else super.transformApply(tree)
+  }
+
+  private def transformTupleCons(tree: tpd.Apply)(implicit ctx: Context): Tree = {
+    val TypeApply(_, headType :: tailType :: Nil) = tree.fun
+    val tail :: head :: Nil = tree.args
```

Comment:
I find it very confusing that the arguments of a dynamicCons are reversed. Can we not leave them in the order head/tail?

Reply:
The order is purely historical. I will reverse them.